qiskit/transpiler/passes/optimization/commutation_analysis.py

```diff
@@ -111,15 +112,21 @@ def _commute(node1, node2, cache):
 
     node1_key = (node1.op.name, str(node1.op.params), str(qarg1))
     node2_key = (node2.op.name, str(node2.op.params), str(qarg2))
+
+    op1 = Operator(node1.op)
+    op2 = Operator(node2.op)
+    if is_diagonal_matrix(op1.data) and is_diagonal_matrix(op2.data):
+        # may be better for short depth wide circuits; use depth criteria?
+        return True
```

Comment:
Does it make sense to cache this too? I'm a bit worried that for circuits with identical gates we'll end up spending a lot of time checking if a matrix is diagonal